src/Shared/ServerInfrastructure/StringUtilities.cs

```diff
@@ -16,401 +15,63 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
 
 internal static class StringUtilities
 {
-    public static unsafe string GetAsciiOrUTF8StringNonNullCharacters(this ReadOnlySpan<byte> span, Encoding defaultEncoding)
+    // Null checks must be done independently of this method (if required)
+    public static string GetAsciiOrUTF8String(this ReadOnlySpan<byte> span, Encoding defaultEncoding)
     {
         if (span.IsEmpty)
         {
             return string.Empty;
         }
 
-        fixed (byte* source = &MemoryMarshal.GetReference(span))
+        var resultString = string.Create(span.Length, span, (destination, spanPtr) =>
```

Comment:
```suggestion
        var resultString = string.Create(span.Length, span, static (destination, source) =>
```
I don't like the name `spanPtr` 😉 as it's not a pointer, etc.

The same on other places (I don't comment on the here).

At least the `static` should be applied to the delegate to avoid unintentional captures.